pandas/core/frame.py

```diff
@@ -5689,6 +5689,108 @@ def _construct_result(self, result) -> "DataFrame":
         out.columns = self.columns
         return out
 
+    @Appender(
+        """
+Returns
+-------
+DataFrame
+    DataFrame that shows the differences stacked side by side.
+
+    The resulting index will be a MultiIndex with 'self' and 'other'
+    stacked alternately at the inner level.
+
+See Also
+--------
+Series.compare : Compare with another Series and show differences.
+
+Notes
+-----
+Matching NaNs will not appear as a difference.
+
+Examples
+--------
+>>> df = pd.DataFrame(
+...     {
+...         "col1": ["a", "a", "b", "b", "a"],
+...         "col2": [1.0, 2.0, 3.0, np.nan, 5.0],
+...         "col3": [1.0, 2.0, 3.0, 4.0, 5.0]
+...     },
+...     columns=["col1", "col2", "col3"],
+... )
+>>> df
+  col1  col2  col3
+0    a   1.0   1.0
+1    a   2.0   2.0
+2    b   3.0   3.0
+3    b   NaN   4.0
+4    a   5.0   5.0
+
+>>> df2 = df.copy()
+>>> df2.loc[0, 'col1'] = 'c'
+>>> df2.loc[2, 'col3'] = 4.0
+>>> df2
+  col1  col2  col3
+0    c   1.0   1.0
+1    a   2.0   2.0
+2    b   3.0   4.0
+3    b   NaN   4.0
+4    a   5.0   5.0
+
+Align the differences on columns
+
+>>> df.compare(df2)
+  col1       col3
+  self other self other
+0    a     c  NaN   NaN
+2  NaN   NaN  3.0   4.0
+
+Stack the differences on rows
+
+>>> df.compare(df2, align_axis=0)
+        col1  col3
+0 self     a   NaN
+  other    c   NaN
+2 self   NaN   3.0
+  other  NaN   4.0
+
+Keep all original rows and columns
+
+>>> df.compare(df2, keep_shape=True)
+  col1       col2       col3
+  self other self other self other
+0    a     c  NaN   NaN  NaN   NaN
+1  NaN   NaN  NaN   NaN  NaN   NaN
+2  NaN   NaN  NaN   NaN  3.0   4.0
+3  NaN   NaN  NaN   NaN  NaN   NaN
+4  NaN   NaN  NaN   NaN  NaN   NaN
+
+Keep all original rows and columns and also all original values
+
+>>> df.compare(df2, keep_shape=True, keep_equal=True)
```

Comment:
does keep_equal make sense w/o keep_shape==True? IOW does it stand on its own? can you add an example of just using it

Reply:
Yes I would say so. It helps identify which one could be the "anomaly". I myself have come across such use cases. 

For example, say I'm looking at the date a company publishes 2 types of their quarterly report.
```python
>>> import pandas as pd
>>> df1 = pd.DataFrame(columns=['Q1_filing', 'Q2_filing', 'Q3_filing', 'Q4_filing'], index=list(range(2010, 2020)))
>>> df1['Q1_filing'] = 'Mar'
>>> df1['Q2_filing'] = 'Jun'
>>> df1['Q3_filing'] = 'Sep'
>>> df1['Q4_filing'] = 'Dec'
>>> df1
     Q1_filing Q2_filing Q3_filing Q4_filing
2010       Mar       Jun       Sep       Dec
2011       Mar       Jun       Sep       Dec
2012       Mar       Jun       Sep       Dec
2013       Mar       Jun       Sep       Dec
2014       Mar       Jun       Sep       Dec
2015       Mar       Jun       Sep       Dec
2016       Mar       Jun       Sep       Dec
2017       Mar       Jun       Sep       Dec
2018       Mar       Jun       Sep       Dec
2019       Mar       Jun       Sep       Dec

>>> df2 = df1.copy()
>>> df2.loc[2015, 'Q1_filing'] = 'Apr'
>>> df2.loc[2016, 'Q2_filing'] = 'Jul'
```

By comparing the two, I can see that the discrepancy is in 2015 and 2016, but I don't know which one deviated from the norm.
```python
>>> df1.compare(df2)
     Q1_filing       Q2_filing
          self other      self other
2015       Mar   Apr       NaN   NaN
2016       NaN   NaN       Jun   Jul
```

The natural thing for me to do now is look at 2015 Q2_filing and 2016 Q1_filing where they agree with each other. (You can of course look at the whole thing but sometimes data is too big and I just want to take a look at the relevant ones first)
```python
>>> df1.compare(df2, keep_equal=True)
     Q1_filing       Q2_filing
          self other      self other
2015       Mar   Apr       Jun   Jun
2016       Mar   Mar       Jun   Jul
```
With this result I know probably something is off for the second type of reports.

Reply:
Have added the example in frame.py. ``keep_equal`` does not stand on its own for `Series` so I did not add anything there.